File: frontend/src/services/employeeService.ts
import api from './api';
import type { Employee, Page } from '../types';

export interface EmployeeParams {
  search?: string;
  departmentId?: number;
  status?: string;
  page?: number;
  size?: number;
  sort?: string;
}

export const employeeService = {
  getAll: (params: EmployeeParams = {}) =>
    api.get<Page<Employee>>('/employees', { params }).then((r) => r.data),

  getById: (id: number) =>
    api.get<Employee>(`/employees/${id}`).then((r) => r.data),

  getByUserId: (userId: number) =>
    api.get<Employee>(`/employees/user/${userId}`).then((r) => r.data),

  create: (data: Employee) =>
    api.post<Employee>('/employees', data).then((r) => r.data),

  update: (id: number, data: Employee) =>
    api.put<Employee>(`/employees/${id}`, data).then((r) => r.data),

  delete: (id: number) =>
    api.delete(`/employees/${id}`).then((r) => r.data),

  /** POST /employees/{id}/account — create + link a login account to this employee */
  assignAccount: (id: number, data: { username: string; password: string; role: string }) =>
    api.post<Employee>(`/employees/${id}/account`, data).then((r) => r.data),
};

File: src/main/java/com/ems/employee_management_system/service/EmployeeService.java
package com.ems.employee_management_system.service;

import com.ems.employee_management_system.dto.AssignAccountRequest;
import com.ems.employee_management_system.dto.EmployeeDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for Employee operations.
 */
public interface EmployeeService {

    /**
     * Retrieves all employees matching search criteria (by name/email), department, and status, with pagination/sorting.
     */
    Page<EmployeeDTO> getAllEmployees(String search, Long departmentId, String status, Pageable pageable);

    /**
     * Retrieves an employee by ID.
     */
    EmployeeDTO getEmployeeById(Long id);

    /**
     * Retrieves an employee by linked user ID.
     */
    EmployeeDTO getEmployeeByUserId(Long userId);

    /**
     * Creates a new employee.
     */
    EmployeeDTO createEmployee(EmployeeDTO employeeDTO);

    /**
     * Updates an existing employee.
     */
    EmployeeDTO updateEmployee(Long id, EmployeeDTO employeeDTO);

    /**
     * Deletes an employee.
     */
    void deleteEmployee(Long id);

    /**
     * Creates a new User account with the given credentials and role,
     * then links it to the employee identified by {@code employeeId}.
     * Throws if the employee already has a linked account.
     */
    EmployeeDTO assignAccount(Long employeeId, AssignAccountRequest request);
}

